lib/galaxy/tools/parameters/basic.py

```diff
@@ -1958,16 +1958,19 @@ def from_json(self, value, trans, other_values=None):
                 elif hasattr(v, "dataset") and v.dataset.state in [galaxy.model.Dataset.states.ERROR, galaxy.model.Dataset.states.DISCARDED]:
                     raise ParameterValueError("the previously selected dataset has entered an unusable state", self.name)
                 elif hasattr(v, "dataset"):
-                    match = dataset_matcher.hda_match(v)
-                    if match and match.implicit_conversion:
-                        v.implicit_conversion = True
+                    match = dataset_matcher.hda_match(v, ensure_visible=False)
+                    if match:
+                        if match.implicit_conversion:
+                            v.implicit_conversion = True
+                    else:
+                        raise ParameterValueError(f"data set ({v.name}) with invalid datatype ({v.ext}) supplied to input dataset parameter (expecting {dataset_matcher.param.extensions})", self.name)
```

Comment:
Maybe we should allow `data` as an extension here for the CWL work that doesn't really use extension and for remote data that doesn't have an extension yet? I guess we've benefited this far from this kind of "loophole".

Reply:
I would have guessed since essentially this function is called https://github.com/galaxyproject/galaxy/blob/9bbc7e7206035cf9a796264f578c5a172b4f8309/lib/galaxy/datatypes/data.py#L841 or is `isinstance` working the other way round? 

If I'm wrong I guess checking for `data` should be done in the matcher, or?